### PR TITLE
WIP: Proposed changes for Admin V4 Login.

### PR DESCRIPTION
**Pull request by alesstimec**
Introduces calls to log in with an access token and client credentials and an additional device flow login.

## Checklist

- [ ] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/

**Jira card:** JUJU-

### CODE REVIEW

File: api/interface.go [hunk 1]
```diff
@@ -230,6 +240,21 @@ func WithDialOpts(newOpts DialOpts) DialOption {
...
+// macaroons, access token or client credentials
+// (client id and secret).
+type LoginParams struct {
+	Tag              names.Tag
+	Password         string
+	Nonce            string
+	Macaroons        []macaroon.Slice
+	AccessToken      string
+	ClientID         string
+	ClientSecret     string
```
Comment (wallyworld) [inline api/interface.go:254]: The LoginParams struct is a useful addition but contains all of these somewhat independent attributes; you are expected to set these but not those etc.
We can consider using an OAuthParams struct to hold these ones. But I think we should consider a separate set of Login methods. This is reinforced by the messy logic in the Login call below.

File: api/state.go [hunk 2]
```diff
@@ -128,24 +221,64 @@ func (st *state) Login(tag names.Tag, password, nonce string, macaroons []macaro
...
+// Login authenticates as the entity with the given name and password
+// or macaroons. Subsequent requests on the state will act as that entity.
+// This method is usually called automatically by Open. The machine nonce
+// should be empty unless logging in as a machine agent.
+func (st *state) Login(p LoginParams) error {
+	var result params.LoginResult
+	var err error
+
+	switch {
+	case p.AccessToken == "" && p.ClientID == "" && p.ClientSecret == "" && p.Password == "" && p.Macaroons == nil:
```
Comment (wallyworld) [inline api/state.go:249]: Following on from previous comment, I think this is messy and we are trying to stuff orthogonal use cases into the one Login call. What about having public login methods for

LoginWithUserCredentials <- the current v3 login
LoginWithClientCredentials
LoginWithDeviceFlow
LoginWithAccessToken

or something

**Review by wallyworld — COMMENTED** [review]
Thanks for starting this. Some comments in line.

**Review by tlm — CHANGES_REQUESTED** [review]
Thank you for the work,

I have had a look at the PR and I don't believe it would be very responsible of us to let this go through in any remote form that it currently looks at.

Some points I would like to make and ask clarification on.
- Are we able to get a detailed explanation as to the why for these changes. It is very hard to review this without understand what we are trying to solve.
- I don't think we should be overloading our login structure with members and concerns from multiple login flows. It's confusing error prone and not easy to reason about in isolation.
- The only login flow that the Juju controller should ever support and by extension JAAS is that of username and password login. All this code smells like we are implementing a M2M workflows for oAuth. If that is the case the we should be doing those login flows with the authorisation server and not a controller. We can't and shouldn't break well defined IETF specifications.
- If the juju client is going to support multiple login mechanisms then I believe these need to be ducked typed out into individual providers of login and the correct one passed to the client so the code doesn't care about how it gets done but that at the end of it has an upgraded web socket that the client can use for api access.
- For passing of "tokens" to the controller I think this should solely happen over the HTTP layer and not the facade layer.
- If we are going to be storing an authorisation token into state at any point it suggests that we have insight into the auth token and can figure out if it is still valid. At this point we need to add retry logic to support getting a new token.
- Is this m2m flow planned for use with the Juju cli or is this just for machine to machine? It is very unclear at the moment.

Happy to offer up my time to help design this in a way that works for client and controller.

**Pull request closed without merging**